frappe/authorizations.py

```diff
@@ -0,0 +1,685 @@
+                    return False
+                else:
+                    result = set(result) & set(auth) if result else set(auth)
+                if not result:
+                    if verbose: print('auth check,individual authorization not in same auth ID', result, auth)
+                    check_log['reason'] = 'not authorized for multi auth fields'
+                    save_check_log(check_log, user, auth_obj_recs, doc)
+                    return False
+        if verbose: print('_auth_check, result:', result)
+        if not bool(result):
+            check_log['reason'] = 'not authorized for multi auth objs'
+            save_check_log(check_log, user)
+    return bool(result)
+
+def save_check_log(check_log, user, auth_obj_recs=None, doc=None):
+    if doc:
+        auth_doc = get_auth_key('', doc, as_list=1)
+    else:
+        auth_doc = [{'auth_obj': rec[0], 'auth_field': rec[1]} for rec in auth_obj_recs or []]
+    check_log['auth_doc'] = auth_doc
+    if auth_obj_recs:
+        check_log['authorization_object'] = auth_obj_recs[0][0]
+        check_log['authorization_field'] = auth_obj_recs[0][1]
+    auth_objs = [rec[0] for rec in auth_obj_recs]
+    auths = [auth for auth in get_user_authorizations(user) or [] if auth[1] in auth_objs]
+    if auths:
+        auths = sorted(auths, key=operator.itemgetter(0))	  # sort by auth_id column for better readability
+        fieldname = ['authorization_id', 'authorization_object', 'authorization_field', 'value_from', 'value_to', 'role']
+        recs = []
+        for auth in (auths or []):
+            rec = {}
+            for i in range(len(auth)):	# extract concatenated auth ID's ID 
+                rec[fieldname[i]] = auth[i].split(':')[-1] if i == 0 else auth[i]
+            recs.append(rec)
+        check_log['authorizations'] = recs
+    auth_check_log = frappe.get_doc(check_log)
+    auth_check_log.flags.ignore_links = True	# to improve performance
+    frappe.local.rollback_observers.append(auth_check_log)
+    old_name = '%s-%s' % (user, check_log.get('doc_type'))
+    frappe.delete_doc('Authorization Check Log', old_name, force=1, ignore_permissions=1, for_reload=1)
+    auth_check_log.insert(ignore_permissions=1)
+    
+def auth_check_doc_fields(doc, user=None):
+    if not user: user = frappe.session.user
+    if user=="Administrator" or frappe.flags.in_install:        
+        return True	
+    auth_objs = get_docfield_auth_objs(doc.doctype)
+    prev_auth_obj = ''
+    result = True
+    auth_check_result = True
+    for (auth_obj, fieldname) in auth_objs:
+        if auth_obj != prev_auth_obj:
+            auth_check_result = auth_check(doc.doctype, 'read', doc, user, auth_obj)
+            prev_auth_obj = auth_obj
+        if not auth_check_result:
+            doc.set(fieldname, None)
+            result = False
+    return result
+
+def get_match_conditions(doctype, act='read', user=None, tables=None, parent_doctype=None, verbose=None):
```

Comment:
function size is too big

Reply:
these days, I am busy with other things, will focus on this topic soon.